### ISSUE TITLE
Pytorch 1.4 cudnn backward issue

### ISSUE THREAD

**Issue by Lewisracing** [pytorch/pytorch#37965]
The Error is ' cudnn RNN backward can only be called in training mode.'

I'm doing a DDPG with LSTM network. Part of the code is shown blow.
It works fine on CPU but when run on GPU, it reports the error.
But! As you can see that I've ready set the critic network to train mode why it still reports error literally four lines after that??

      def learn(self):

        if self.memory.mem_cntr<self.batch_size:
            return

        state,action,reward,new_state,done=self.memory.sample_buffer(self.batch_size)
        reward=T.tensor(reward,dtype=T.float).to(self.critic.device)
        done=T.tensor(done).to(self.critic.device)
        new_state=T.tensor(new_state,dtype=T.float).to(self.critic.device).view((-1,1,12,16))

        action=T.tensor(action,dtype=T.float).to(self.critic.device)
        state=T.tensor(state,dtype=T.float).to(self.critic.device)

        self.target_actor.eval()
        self.target_critic.eval()
        self.critic.eval()

        target_actions=self.target_actor.forward(new_state)
        critic_value_=self.target_critic.forward(new_state,target_actions)
        critic_value=self.critic.forward(state,action)

        target=[]
        for j in range(self.batch_size):
            target.append(reward[j]+self.gamma*critic_value_[j]*done[j])
        target=T.tensor(target).to(self.critic.device)
        target=target.view(self.batch_size,1)

        self.critic.train()
        
        self.critic.optimizer.zero_grad()
        critic_loss=F.mse_loss(target,critic_value)
        critic_loss.backward()
        self.critic.optimizer.step()

        self.critic.eval()
        self.actor.optimizer.zero_grad()
        mu=self.actor.forward(state)
        self.actor.train()
        actor_loss=-self.critic.forward(state,mu)
        actor_loss=T.mean(actor_loss)
        actor_loss.backward()
        self.actor.optimizer.step()

        self.update_network_parameters()

I checked the .training boolean value and confirmed that all my networks are in training mode. Yet it still reports the issue

cc @zou3519

**Comment by SsnL** [comment]
because the computation graph was created before you call `.train`

**Comment by Lewisracing** [comment]
> because the computation graph was created before you call `.train`

So there’s literally no way to overcome this…? I’ve no idea

**Comment by SsnL** [comment]
just call .train before you create the graph......

**Comment by agolynski** [comment]
@Lewisracing did it help?

**Comment by agolynski** [comment]
@Lewisracing i see. do you mind me asking what is the purpose of 

   self.target_actor.eval()
    self.target_critic.eval()
    self.critic.eval()

in the beginning of your code? are you done training model at this point and then you restart training again?

**Comment by Lewisracing** [comment]
@agolynski  The is part of the 'learning step' of the DDPG reinforcement learning where in order to calculate the critic NN loss, I have to first get the prediction value from the 'target_actor','target_critic','critic' networks. I've re-edited the issue so you can see the full learn function on the top

**Comment by agolynski** [comment]
i am not very familiar with DDPG reinforcement learning  specifics, do you need to enter eval mode in order to get prediction values? does any of your underlying modules get affected by train/eval mode, e.g. dropout or batchednorm, see https://jamesmccaffrey.wordpress.com/2019/01/23/pytorch-train-vs-eval-mode/

**Comment by Lewisracing** [comment]
@agolynski I'm afraid yes...  I have batchnorms in those networks

**Comment by albanD** [comment]
Hi,

This is a limitation of the `cudnn` implementation I'm afraid so there isn't much you can do with cudnn.
The solution is to disable cudnn for this part of your network and use pytorch's implementation that support backward in eval mode:
```
# some forward code

with torch.backends.cudnn.flags(enabled=False):
    out = problematic_module(inputs)

# more forward code
```

**Comment by Lewisracing** [comment]
Alright then.. Thanks for all the help
